File: Striver/Trees/Q08.py
# Iterative Postorder Traversal of Binary Tree using 1 Stack
from Node import Node

def postOrder(root):
    ans = []
    if root is None:
        return ans
    
    curr = root
    stack = []

    while curr is not None or len(stack) > 0:
        if curr is not None:
            stack.append(curr)
            curr = curr.left
        else:
            temp = stack[-1].right
            if temp is None:
                temp = stack.pop()
                ans.append(temp.data)
                while len(stack) > 0 and temp == stack[-1].right:
                    temp = stack.pop()
                    ans.append(temp.data)
            else:
                curr = temp

    return ans

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

print(postOrder(root))

File: Striver/Trees/Q08.java
package Striver.Trees;
import java.util.*;

public class Q08 {
    // Iterative Postorder Traversal using 1 Stack

    public static List<Integer> postorderTraversal1(Node root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;

        Stack<Node> st = new Stack<>();
        st.push(root);

        while (!st.isEmpty()) {
            Node curr = st.pop();
            ans.add(curr.data);

            if (curr.left != null) st.push(curr.left);
            if (curr.right != null) st.push(curr.right);
        }

        Collections.reverse(ans);
        return ans;
    }

    public static List<Integer> postorderTraversal2(Node root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;

        Stack<Node> st = new Stack<>();
        st.push(root);
        Node curr = root;

        while (curr != null || !st.isEmpty()) {
            if (curr != null) {
                st.push(curr);
                curr = curr.left;
            }
            else{
                Node temp = st.peek().right;
                if (temp == null) {
                    temp = st.pop();
                    ans.add(temp.data);

                    while (!st.isEmpty() && temp == st.peek().right) {
                        temp = st.pop();
                        ans.add(temp.data);
                    }
                }
                else curr = temp;
            }
        }

        return ans;
    }

    public static void main(String[] args) {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.right.left = new Node(6);
        root.right.right = new Node(7);

        System.out.println(postorderTraversal1(root));
        System.out.println(postorderTraversal2(root));
    }
}
